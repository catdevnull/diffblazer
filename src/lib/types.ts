export type DiffClassNames = {
	del: string
	ins: string
	replace: string
	mod: string
}

export type DiffTags = {
	del: string
	ins: string
}

export type HTMLDiffOptions = {
	/**
	 * This value defines balance between speed and memory utilization. The higher it is the faster it works and more memory it consumes.
	 *
	 * A bigger value won't always increase speed, and most probably will
	 * negatively affect performance.
	 *
	 * @status experimental
	 * @default 4
	 */
	matchGranularity: number
	repeatingTokensAccuracy: number
	ignoreWhiteSpaceDifferences: boolean
	orphanMatchThreshold: number
	atomicTags: string[]
	classNames: DiffClassNames
	tags: DiffTags
}
