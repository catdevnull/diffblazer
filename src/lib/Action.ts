enum Action {
	equal,
	delete,
	insert,
	replace,
}

export default Action
