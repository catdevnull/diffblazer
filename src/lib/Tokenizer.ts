import { Tokenizer } from 'htmlparser2'

export type TextToken = { type: 'text'; value: string; length: number }
export type TagToken = {
	type: 'tag-start' | 'tag-end' | 'tag-full'
	name: string
	raw: string
	length: number
}
export type Token = TextToken | TagToken

type TagRange = { start?: number; end?: number; nameEnd?: number }
const DEFAULT_ATOMIC_TAGS = ['head', 'style', 'script', 'math', 'object', 'video', 'iframe']
/**
 * Tokenizes a string of HTML.
 **/
function tokenizeHtml(text: string, atomicTags: string[] = DEFAULT_ATOMIC_TAGS) {
	const tokens: Token[] = []
	let tagRange: TagRange = {}

	// we don't actually ignore these tags but only split their
	// contents as a single token instead. This way we can be
	// performant without adding special conditions.
	let ignoring: { tag: string; content: string } | undefined = undefined

	const tokenizer = new Tokenizer(
		{ decodeEntities: false, xmlMode: false },
		{
			onclosetag(start, end) {
				pushTag({ start, end }, 'closing')
			},
			ondeclaration(start, end) {
				pushTag({ start, end }, 'declaration')
			},
			onopentagend(end) {
				tagRange.end = end
				pushTag(tagRange)
			},
			onopentagname(start, end) {
				tagRange.start = start
				tagRange.nameEnd = end
			},
			onselfclosingtag(end) {
				tagRange.end = end
				pushTag(tagRange, 'self-closing')
			},
			ontext(start, end) {
				if (ignoring) {
					ignoring.content += text.slice(start, end)
					return
				}

				const tokenized = text.slice(start, end).split(/(\s+|&.+?;|[^A-Za-z0-9#])/gm)
				tokenized.forEach((t) => {
					if (!t) return
					tokens.push({ type: 'text', value: t, length: t.length })
				})
			},

			onprocessinginstruction() {},
			onattribdata() {},
			onattribend() {},
			onattribentity() {},
			onattribname() {},
			oncdata() {},
			oncomment() {},
			onend() {},
			ontextentity() {},
		},
	)
	tokenizer.write(text)
	tokenizer.end()

	function pushTag(range: TagRange, type?: 'closing' | 'declaration' | 'self-closing') {
		if (!range.start || !range.end) {
			tagRange = {}
			return
		}

		const tagName = text.slice(range.start, range.nameEnd || range.end)
		const tag = range.nameEnd ? text.slice(range.start, range.end) : tagName
		let tagStart = type === 'declaration' ? `<!` : type === 'closing' ? '</' : '<'
		const raw = `${tagStart}${tag}>`

		if (ignoring || atomicTags.includes(tagName)) {
			ignoring = ignoring || {
				tag: tagName,
				content: '',
			}
			ignoring.content += raw

			if (type && ignoring.tag === tagName) {
				tokens.push({
					type: 'tag-full',
					name: ignoring.tag,
					raw: ignoring.content,
					length: ignoring.content.length,
				})
			}
		} else {
			tokens.push({
				raw,
				type: type === 'closing' ? 'tag-end' : 'tag-start',
				name: tagName,
				length: range.end - range.start,
			})
		}
		tagRange = {}
	}

	return tokens
}

function joinTokens(tokens: Token[], separator?: string): string {
	let str = ''
	for (let i = 0; i < tokens.length; ++i) {
		const token = tokens[i]
		str += token.type === 'text' ? token.value : token.raw
		if (separator && i < tokens.length - 1) str += separator
	}
	return str
}

export { joinTokens, tokenizeHtml, DEFAULT_ATOMIC_TAGS }
