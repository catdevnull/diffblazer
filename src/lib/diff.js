function HtmlDiff() {
	console.log('Test Function')
}

export default HtmlDiff
