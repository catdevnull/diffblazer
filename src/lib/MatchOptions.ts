import { TagToken } from './Tokenizer'

type MatchOptions = {
	blockSize: number
	ignoreWhitespaceDifferences: boolean
	matchers: Record<string, (token: TagToken) => string>
}

const MatchOptions: MatchOptions = {
	matchers: {},
	blockSize: 0,
	ignoreWhitespaceDifferences: false,
}

export default MatchOptions
