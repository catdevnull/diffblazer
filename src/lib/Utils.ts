import * as WordSplitter from './Tokenizer'

const whitespaceRegex = /^(\s|&nbsp;)+$/

function isTag(token: WordSplitter.Token) {
	return token.type !== 'text'
}

function isNotTag(token: WordSplitter.Token): token is WordSplitter.TextToken {
	return !isTag(token)
}

function wrapText(text: string, tagName: string, cssClass: string) {
	return ['<', tagName, " class='", cssClass, "'>", text, '</', tagName, '>'].join('')
}

function isWhiteSpace(character: string) {
	return whitespaceRegex.test(character)
}

export { isNotTag, isTag, wrapText, isWhiteSpace }
