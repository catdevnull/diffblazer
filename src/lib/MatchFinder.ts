import Match from './Match'
import MatchOptions from './MatchOptions'
import * as Utils from './Utils'
import * as Tokenizer from './Tokenizer'

function putNewToken(block: string[], token: string, blockSize: number) {
	block.push(token)

	if (block.length > blockSize) {
		block.shift()
	}

	if (block.length !== blockSize) {
		return null
	}

	return block.join('')
}

// Finds the longest match in given texts. It uses indexing with fixed granularity that is used to compare blocks of text.
export default class MatchFinder {
	private oldTokens: Tokenizer.Token[]
	private newTokens: Tokenizer.Token[]
	private startInOld: number
	private endInOld: number
	private startInNew: number
	private endInNew: number
	private options: MatchOptions
	private tokenIndices: Map<string, number[]>

	constructor(
		oldTokens: Tokenizer.Token[],
		newTokens: Tokenizer.Token[],
		startInOld: number,
		endInOld: number,
		startInNew: number,
		endInNew: number,
		options: MatchOptions,
	) {
		this.oldTokens = oldTokens
		this.newTokens = newTokens
		this.startInOld = startInOld
		this.endInOld = endInOld
		this.startInNew = startInNew
		this.endInNew = endInNew
		this.options = options
		this.tokenIndices = new Map()
	}

	indexNewTokens() {
		this.tokenIndices = new Map()
		let block: string[] = [''].slice(1)
		let threshold = this.newTokens.length + this.options.repeatingTokensAccuracy
		let blacklist = new Set<string>()

		for (let i = this.startInNew; i < this.endInNew; i++) {
			let token = this.normalizeForIndex(this.newTokens[i])
			let key = putNewToken(block, token, this.options.blockSize)

			if (key === null || blacklist.has(key)) {
				continue
			}

			const index = this.tokenIndices.get(key)
			if (index) {
				// This removes & blacklists tokens that occur too many times.
				// This way it reduces total count of comparison operations
				// and as result the diff algorithm takes less time. But the
				// side effect is that it may detect false differences of
				// the repeating tokens.
				if (index.length >= threshold) {
					blacklist.add(key)
					this.tokenIndices.delete(key)
					continue
				}

				index.push(i)
			} else {
				this.tokenIndices.set(key, [i])
			}
		}
	}

	// Converts the token to index-friendly value so it can be compared with other similar tokens
	normalizeForIndex(token: Tokenizer.Token) {
		if (this.options.ignoreWhitespaceDifferences && token.type === 'text' && Utils.isWhiteSpace(token.value)) {
			return ' '
		}

		return token.type === 'text'
			? token.value
			: this.options.matchers[token.name]
			? this.options.matchers[token.name](token)
			: token.name
	}

	findMatch() {
		this.indexNewTokens()

		if (this.tokenIndices.size === 0) {
			return null
		}

		let bestMatchInOld = this.startInOld
		let bestMatchInNew = this.startInNew
		let bestMatchSize = 0

		let matchLengthAt = new Map()
		const blockSize = this.options.blockSize
		let block: string[] = [''].slice(1)

		for (let indexInOld = this.startInOld; indexInOld < this.endInOld; indexInOld++) {
			let token = this.normalizeForIndex(this.oldTokens[indexInOld])
			let index = putNewToken(block, token, blockSize)

			if (index === null) {
				continue
			}

			let newMatchLengthAt = new Map()

			const tokenIndices = this.tokenIndices.get(index)
			if (!tokenIndices) {
				matchLengthAt = newMatchLengthAt
				continue
			}

			for (let indexInNew of tokenIndices) {
				let newMatchLength = (matchLengthAt.get(indexInNew - 1) || 0) + 1
				newMatchLengthAt.set(indexInNew, newMatchLength)

				if (newMatchLength > bestMatchSize) {
					bestMatchInOld = indexInOld - newMatchLength - blockSize + 2
					bestMatchInNew = indexInNew - newMatchLength - blockSize + 2
					bestMatchSize = newMatchLength
				}
			}

			matchLengthAt = newMatchLengthAt
		}

		return bestMatchSize !== 0 ? new Match(bestMatchInOld, bestMatchInNew, bestMatchSize + blockSize - 1) : null
	}
}
