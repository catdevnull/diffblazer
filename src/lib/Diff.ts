import Action from './Action'
import Match from './Match'
import MatchFinder from './MatchFinder'
import Operation from './Operation'
import MatchOptions from './MatchOptions'
import * as Tokenizer from './Tokenizer'
import * as Utils from './Utils'
import type { DiffClassNames, DiffTags, HTMLDiffOptions } from './types'

const MatchGranuarityMaximum = 4

const specialCaseTags = new Set(['strong', 'em', 'b', 'i', 'big', 'small', 'u', 'sub', 'strike', 's', 'dfn', 'span'])

const ATTRIBUTE_REGEX = (attr: string) => new RegExp(` ${attr}=(.+?['"])`)
const SRC_REGEX = ATTRIBUTE_REGEX('src')
const DATA_REGEX = ATTRIBUTE_REGEX('data')
const MATH_CONTENT_REGEX = /<math.+?>(.+)?<\/math>/

function attributeMatcher(token: Tokenizer.TagToken, attributes: Record<string, RegExp>) {
	let text = token.name
	for (const attr in attributes) {
		if (token.raw.includes(` ${attr}=`)) {
			const regex = attributes[attr]
			const matches = regex.exec(token.raw)
			if (!matches) continue
			text += matches[1]
		}
	}
	return text
}

const matchers = {
	img: (token: Tokenizer.TagToken) => attributeMatcher(token, { src: SRC_REGEX }),
	object: (token: Tokenizer.TagToken) => attributeMatcher(token, { data: DATA_REGEX }),
	math: (token: Tokenizer.TagToken) => {
		const matches = MATH_CONTENT_REGEX.exec(token.raw)
		return matches ? matches[1] : token.name
	},
	video: (token: Tokenizer.TagToken) => attributeMatcher(token, { src: SRC_REGEX }),
	iframe: (token: Tokenizer.TagToken) => attributeMatcher(token, { src: SRC_REGEX }),
}

class HtmlDiff {
	private content: string[]
	private newText: string
	private oldText: string

	private specialTagDiffStack: Tokenizer.TagToken[]
	private newTokens: Tokenizer.Token[]
	private oldTokens: Tokenizer.Token[]
	private matchGranularity: number

	private repeatingTokensAccuracy: number
	private ignoreWhiteSpaceDifferences: boolean
	private orphanMatchThreshold: number
	private atomicTags: string[]
	private classNames: DiffClassNames
	private tags: DiffTags

	constructor(oldText: string, newText: string, options?: Partial<HTMLDiffOptions>) {
		this.content = []
		this.newText = newText
		this.oldText = oldText

		this.specialTagDiffStack = []
		this.newTokens = []
		this.oldTokens = []
		this.matchGranularity = options?.matchGranularity ?? MatchGranuarityMaximum

		this.repeatingTokensAccuracy = options?.repeatingTokensAccuracy ?? 1.0
		this.ignoreWhiteSpaceDifferences = options?.ignoreWhiteSpaceDifferences ?? false
		this.orphanMatchThreshold = options?.orphanMatchThreshold ?? 0.0
		this.atomicTags = options?.atomicTags ?? Tokenizer.DEFAULT_ATOMIC_TAGS
		this.classNames = options?.classNames ?? {
			del: 'diffdel',
			ins: 'diffins',
			replace: 'diffmod',
			mod: 'mod',
		}
		this.tags = options?.tags ?? {
			del: 'del',
			ins: 'ins',
		}
	}

	static execute(oldText: string, newText: string, options?: Partial<HTMLDiffOptions>) {
		return new HtmlDiff(oldText, newText, options).build()
	}

	build() {
		if (this.oldText === this.newText) {
			return this.newText
		}

		this.tokenizeInputs()

		this.matchGranularity = Math.min(this.matchGranularity, this.oldTokens.length, this.newTokens.length)
		let operations = this.operations()
		for (let item of operations) {
			this.performOperation(item)
		}

		return this.content.join('')
	}

	tokenizeInputs() {
		this.oldTokens = Tokenizer.tokenizeHtml(this.oldText, this.atomicTags)

		//free memory, allow it for GC
		this.oldText = ''

		this.newTokens = Tokenizer.tokenizeHtml(this.newText, this.atomicTags)

		//free memory, allow it for GC
		this.newText = ''
	}

	performOperation(opp: Operation) {
		switch (opp.action) {
			case Action.equal:
				this.processEqualOperation(opp)
				break
			case Action.delete:
				this.processDeleteOperation(opp, this.classNames.del)
				break
			case Action.insert:
				this.processInsertOperation(opp, this.classNames.ins)
				break
			case Action.none:
				break
			case Action.replace:
				this.processReplaceOperation(opp)
				break
		}
	}

	processReplaceOperation(opp: Operation) {
		this.processDeleteOperation(opp, this.classNames.replace)
		this.processInsertOperation(opp, this.classNames.replace)
	}

	processInsertOperation(opp: Operation, cssClass: string) {
		let text = this.newTokens.filter((_, pos) => pos >= opp.startInNew && pos < opp.endInNew)
		this.insertTag(this.tags.ins, cssClass, text)
	}

	processDeleteOperation(opp: Operation, cssClass: string) {
		let text = this.oldTokens.filter((_, pos) => pos >= opp.startInOld && pos < opp.endInOld)
		this.insertTag(this.tags.del, cssClass, text)
	}

	processEqualOperation(opp: Operation) {
		let result = this.newTokens.filter((_, pos) => pos >= opp.startInNew && pos < opp.endInNew)
		this.content.push(Tokenizer.joinTokens(result))
	}

	insertTag(tag: string, cssClass: string, tokens: Tokenizer.Token[]) {
		while (tokens.length) {
			let nonTags = this.extractConsecutiveTokens(
				tokens,
				(x) => Utils.isNotTag(x) || this.atomicTags.includes(x.name) || x.name === 'img',
			)

			let specialCaseTagInjection = ''
			let specialCaseTagInjectionIsBefore = false

			if (nonTags.length !== 0) {
				let text = Utils.wrapText(Tokenizer.joinTokens(nonTags), tag, cssClass)
				this.content.push(text)
			} else {
				const firstToken = tokens[0]
				if (firstToken.type === 'tag-start' && specialCaseTags.has(firstToken.name)) {
					this.specialTagDiffStack.push(firstToken)
					specialCaseTagInjection = `<${this.tags.ins} class='${this.classNames.mod}'>`

					if (tag === this.tags.del) {
						tokens.shift()
						while (tokens.length > 0 && tokens[0].type === 'tag-start' && specialCaseTags.has(tokens[0].name)) {
							tokens.shift()
						}
					}
				} else if (firstToken.type === 'tag-end' && specialCaseTags.has(firstToken.name)) {
					let openingTag = this.specialTagDiffStack.length === 0 ? null : this.specialTagDiffStack.pop()

					const lastToken = tokens.at(-1)
					const openingAndClosingTagsMatch =
						openingTag && lastToken?.type === 'tag-end' && openingTag.name === lastToken.name

					if (openingAndClosingTagsMatch) {
						specialCaseTagInjection = `</${this.tags.ins}>`
						specialCaseTagInjectionIsBefore = true
					}

					if (tag === this.tags.del) {
						tokens.shift()
						while (tokens.length > 0 && tokens[0].type === 'tag-end' && specialCaseTags.has(tokens[0].name)) {
							tokens.shift()
						}
					}
				}

				if (tokens.length === 0 && specialCaseTagInjection.length === 0) {
					break
				}

				if (specialCaseTagInjectionIsBefore) {
					this.content.push(
						specialCaseTagInjection + Tokenizer.joinTokens(this.extractConsecutiveTokens(tokens, Utils.isTag)),
					)
				} else {
					this.content.push(
						Tokenizer.joinTokens(this.extractConsecutiveTokens(tokens, Utils.isTag)) + specialCaseTagInjection,
					)
				}
			}
		}
	}

	extractConsecutiveTokens(tokens: Tokenizer.Token[], condition: (value: Tokenizer.Token) => boolean) {
		let indexOfFirstTag: number | null = null

		for (let i = 0; i < tokens.length; i++) {
			let token = tokens[i]

			if (i === 0 && token.type === 'text' && token.value === ' ') {
				token.value = '&nbsp;'
			}

			if (!condition(token)) {
				indexOfFirstTag = i
				break
			}
		}

		if (indexOfFirstTag !== null) {
			let items = tokens.filter((_, pos) => pos >= 0 && pos < indexOfFirstTag!)
			if (indexOfFirstTag > 0) {
				tokens.splice(0, indexOfFirstTag)
			}

			return items
		} else {
			let items = tokens.filter((_, pos) => pos >= 0 && pos < tokens.length)
			tokens.splice(0, tokens.length)
			return items
		}
	}

	operations() {
		let positionInOld = 0
		let positionInNew = 0
		let operations: Operation[] = []

		let matches = this.matchingBlocks()
		matches.push(new Match(this.oldTokens.length, this.newTokens.length, 0))

		let matchesWithoutOrphans = this.removeOrphans(matches)

		for (let match of matchesWithoutOrphans) {
			if (match === null) continue
			let matchStartsAtCurrentPositionInOld = positionInOld === match.startInOld
			let matchStartsAtCurrentPositionInNew = positionInNew === match.startInNew

			let action

			if (!matchStartsAtCurrentPositionInOld && !matchStartsAtCurrentPositionInNew) {
				action = Action.replace
			} else if (matchStartsAtCurrentPositionInOld && !matchStartsAtCurrentPositionInNew) {
				action = Action.insert
			} else if (!matchStartsAtCurrentPositionInOld) {
				action = Action.delete
			} else {
				action = Action.none
			}

			if (action !== Action.none) {
				const op = new Operation(action, positionInOld, match.startInOld, positionInNew, match.startInNew)
				operations.push(op)
			}

			if (match.size !== 0) {
				operations.push(new Operation(Action.equal, match.startInOld, match.endInOld, match.startInNew, match.endInNew))
			}

			positionInOld = match.endInOld
			positionInNew = match.endInNew
		}

		return operations
	}

	*removeOrphans(matches: Match[]) {
		let prev = null
		let curr = null

		for (let next of matches) {
			if (curr === null) {
				prev = new Match(0, 0, 0)
				curr = next
				continue
			}

			if (
				(prev?.endInOld === curr.startInOld && prev?.endInNew === curr.startInNew) ||
				(curr.endInOld === next.startInOld && curr.endInNew === next.startInNew)
			) {
				yield curr
				curr = next
				continue
			}

			let sumLength = (t: number, n: Tokenizer.Token) => t + n.length

			let oldDistanceInChars = this.oldTokens.slice(prev?.endInOld, next.startInOld).reduce(sumLength, 0)
			let newDistanceInChars = this.newTokens.slice(prev?.endInNew, next.startInNew).reduce(sumLength, 0)
			let currMatchLengthInChars = this.newTokens.slice(curr.startInNew, curr.endInNew).reduce(sumLength, 0)
			if (currMatchLengthInChars > Math.max(oldDistanceInChars, newDistanceInChars) * this.orphanMatchThreshold) {
				yield curr
			}

			prev = curr
			curr = next
		}

		yield curr
	}

	matchingBlocks() {
		let matchingBlocks: Match[] = []
		this.findMatchingBlocks(0, this.oldTokens.length, 0, this.newTokens.length, matchingBlocks)
		return matchingBlocks
	}

	findMatchingBlocks(
		startInOld: number,
		endInOld: number,
		startInNew: number,
		endInNew: number,
		matchingBlocks: Match[],
	) {
		let match = this.findMatch(startInOld, endInOld, startInNew, endInNew)

		if (match !== null) {
			if (startInOld < match.startInOld && startInNew < match.startInNew) {
				this.findMatchingBlocks(startInOld, match.startInOld, startInNew, match.startInNew, matchingBlocks)
			}

			matchingBlocks.push(match)

			if (match.endInOld < endInOld && match.endInNew < endInNew) {
				this.findMatchingBlocks(match.endInOld, endInOld, match.endInNew, endInNew, matchingBlocks)
			}
		}
	}

	findMatch(startInOld: number, endInOld: number, startInNew: number, endInNew: number) {
		for (let i = this.matchGranularity; i > 0; i--) {
			let options = MatchOptions
			options.blockSize = i
			options.repeatingTokensAccuracy = this.repeatingTokensAccuracy
			options.ignoreWhitespaceDifferences = this.ignoreWhiteSpaceDifferences
			options.matchers = matchers

			let finder = new MatchFinder(this.oldTokens, this.newTokens, startInOld, endInOld, startInNew, endInNew, options)
			let match = finder.findMatch()
			if (match !== null) {
				return match
			}
		}

		return null
	}

	// private expandToWrapTag(tokens: Tokenizer.Token[], start: number, end: number) {
	// 	let newEnd = end
	// 	for (let i = start; i < end; ++i) {
	// 		const token = tokens[i]
	// 		if (token.type === 'tag-start' && !token.selfClosing) {
	// 			let endToken = tokens.findIndex((t, index) => index > i && t.type === 'tag-end' && t.name === token.name)
	// 			if (endToken === -1) continue
	// 			endToken++
	// 			if (endToken <= newEnd) continue
	// 			newEnd = endToken
	// 		}
	// 	}
	// 	return newEnd
	// }
}

export default HtmlDiff
