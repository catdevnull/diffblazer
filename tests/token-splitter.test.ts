import { expect, it, describe } from 'vitest'
import { tokenizeHtml } from '../src/lib/Tokenizer'

describe('tokenize html', function () {
	describe('when called with text', function () {
		it('should return 7', function () {
			expect(tokenizeHtml('this is a test').length).to.equal(7)
		})
	})

	describe('when called with html', function () {
		const html = '<p>this is a <strong>test</strong></p>'

		it('should return 11', function () {
			expect(tokenizeHtml(html).length).to.equal(11)
		})

		it('should remove any html comments', function () {
			const html = '<p> this is <!-- a comment! --> </p>'
			expect(tokenizeHtml(html).length).to.equal(8)
		})
	})

	it('should identify contiguous whitespace as a single token', function () {
		expect(tokenizeHtml('a           b')).toMatchSnapshot()
	})

	it('should identify a single space as a single token', function () {
		expect(tokenizeHtml(' a b ')).toMatchSnapshot()
	})

	it('should identify self closing tags as tokens', function () {
		expect(tokenizeHtml('<p>hello</br>goodbye</p>')).toMatchSnapshot()
	})

	describe('when encountering atomic tags', function () {
		it('should identify an image tag as a single token', function () {
			expect(tokenizeHtml('<p><img src="1.jpg"><img src="2.jpg"></p>', ['img'])).toMatchSnapshot()
		})

		it('should identify an iframe tag as a single token', function () {
			expect(tokenizeHtml('<p><iframe src="sample.html"></iframe></p>')).toMatchSnapshot()
		})

		it('should identify an object tag as a single token', function () {
			expect(tokenizeHtml('<p><object><param name="1" /><param name="2" /></object></p>')).toMatchSnapshot()
		})

		it('should identify a math tag as a single token', function () {
			expect(
				tokenizeHtml(
					'<p><math xmlns="http://www.w3.org/1998/Math/MathML">' +
						'<mi>&#x03C0;<!-- Ï€ --></mi>' +
						'<mo>&#x2062;<!-- &InvisibleTimes; --></mo>' +
						'<msup><mi>r</mi><mn>2</mn></msup></math></p>',
				),
			).toMatchSnapshot()
		})

		it('should identify an svg tag as a single token', function () {
			expect(
				tokenizeHtml(
					'<p><svg width="100" height="100">' +
						'<circle cx="50" cy="50" r="40" stroke="green" stroke-width="4" ' +
						'fill="yellow" /></svg></p>',
				),
			).toMatchSnapshot()
		})

		it('should identify a script tag as a single token', function () {
			expect(tokenizeHtml('<p><script>console.log("hi");</script></p>')).toMatchSnapshot()
		})

		it('should identify style tag as a single token', () => {
			expect(tokenizeHtml('<style>hello what is this new tag # . ! ( ) word </style>')).toMatchSnapshot()
		})

		it('should identify head tag as a single token', () => {
			expect(
				tokenizeHtml(
					`<head>
						<title>hello</title>
						<script>hello what is this new tag # . ! ( ) word </script>
						<style>hello what is this new tag # . ! ( ) word </style>
					</head>`,
				),
			).toMatchSnapshot()
		})
	})
})
