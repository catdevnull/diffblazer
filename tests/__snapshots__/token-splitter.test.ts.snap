// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tokenize html > should identify a single space as a single token 1`] = `
[
  {
    "length": 1,
    "type": "text",
    "value": " ",
  },
  {
    "length": 1,
    "type": "text",
    "value": "a",
  },
  {
    "length": 1,
    "type": "text",
    "value": " ",
  },
  {
    "length": 1,
    "type": "text",
    "value": "b",
  },
  {
    "length": 1,
    "type": "text",
    "value": " ",
  },
]
`;

exports[`tokenize html > should identify contiguous whitespace as a single token 1`] = `
[
  {
    "length": 1,
    "type": "text",
    "value": "a",
  },
  {
    "length": 11,
    "type": "text",
    "value": "           ",
  },
  {
    "length": 1,
    "type": "text",
    "value": "b",
  },
]
`;

exports[`tokenize html > should identify doctype 1`] = `
[
  {
    "length": 7,
    "name": "doctype",
    "raw": "<!doctype>",
    "type": "tag-start",
  },
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 5,
    "type": "text",
    "value": "hello",
  },
  {
    "length": 1,
    "name": "p",
    "raw": "</p>",
    "type": "tag-end",
  },
]
`;

exports[`tokenize html > should identify self closing tags as tokens 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 5,
    "type": "text",
    "value": "hello",
  },
  {
    "length": 2,
    "name": "br",
    "raw": "</br>",
    "type": "tag-end",
  },
  {
    "length": 7,
    "type": "text",
    "value": "goodbye",
  },
  {
    "length": 1,
    "name": "p",
    "raw": "</p>",
    "type": "tag-end",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify a math tag as a single token 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 123,
    "name": "math",
    "raw": "<math xmlns="http://www.w3.org/1998/Math/MathML"><mi>&#x03C0;</mi><mo>&#x2062;</mo><msup><mi>r</mi><mn>2</mn></msup></math>",
    "type": "tag-full",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify a script tag as a single token 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 35,
    "name": "script",
    "raw": "<script>console.log("hi");</script>",
    "type": "tag-full",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify an iframe tag as a single token 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 35,
    "name": "iframe",
    "raw": "<iframe src="sample.html"></iframe>",
    "type": "tag-full",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify an image tag as a single token 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify an object tag as a single token 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 53,
    "name": "object",
    "raw": "<object><param name="1" /><param name="2" /></object>",
    "type": "tag-full",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify an svg tag as a single token 1`] = `
[
  {
    "length": 1,
    "name": "p",
    "raw": "<p>",
    "type": "tag-start",
  },
  {
    "length": 28,
    "name": "svg",
    "raw": "<svg width="100" height="100">",
    "type": "tag-start",
  },
  {
    "length": 77,
    "name": "circle",
    "raw": "<circle cx="50" cy="50" r="40" stroke="green" stroke-width="4" fill="yellow" />",
    "type": "tag-start",
  },
  {
    "length": 3,
    "name": "svg",
    "raw": "</svg>",
    "type": "tag-end",
  },
  {
    "length": 1,
    "name": "p",
    "raw": "</p>",
    "type": "tag-end",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify head tag as a single token 1`] = `
[
  {
    "length": 176,
    "name": "head",
    "raw": "<head>
						<title>hello</title>
						<script>hello what is this new tag # . ! ( ) word </script>
						<style>hello what is this new tag # . ! ( ) word </style>
					</head>",
    "type": "tag-full",
  },
]
`;

exports[`tokenize html > when encountering atomic tags > should identify style tag as a single token 1`] = `
[
  {
    "length": 57,
    "name": "style",
    "raw": "<style>hello what is this new tag # . ! ( ) word </style>",
    "type": "tag-full",
  },
]
`;
